llvm/docs/LangRef.rst

```diff
@@ -23091,6 +23091,1189 @@ Returns another pointer that aliases its argument but which has no associated
 ``invariant.group`` metadata.
 It does not read any memory and can be speculated.
 
+.. _fpbuiltin:
+
+Floating-Point Builtin Intrinsics
+-------------------------------------
+
+These intrinsics are used to represent common floating-point operations with
+the explicit expectation that the semantics of the operation may be modified
+by call-site attributes that are specific to these intrinsics. Although many
+of these operations correspond directly to functions defined by the standard
+C math library, these intrinsics are intended to allow replacement of the
+intrinsic with implementation outside the standard library, such as vector
+implementations of the operation or alternate implementations to satisfy
+different accuracy requirements.
+
+The following call-site attributes are currently recognized as being associated
+with the floating-point builtin intrinsics:
+
+``"fpbuiltin-max-error"="<ulp>"``
+    This attribute specifies the required accuracy for the operation in ULPs.
+    The accuracy value must be a non-negative floating-point number. A value
+    of 0.5 or less indicates that the result is required to be correctly
+    rounded according to IEEE-754 rules. The default rounding mode
+    (round-to-nearest) may be assumed.
+
+    If this attribute is absent, basic operations (fadd, fsub, fmul, fdiv,
+    frem, and sqrt) are assumed to provide correctly rounded result. The
+    accuracy of other operations is target-dependent, corresponding to the
+    accuracy of the target-default implementation of the operation (usually
+    the implementation provided by the standard math library). If this
+    attribute is present, the intrinsic may only be replaced with
+    implementations which are known to provide at least the accuracy described.
+    An implementation which is more accurate than required by this attribute
+    may be used.
+
```

Comment:
Can we add explicit rounding modes as attributes? This would be to support CUDA/HIP intrinsics inputs: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#intrinsic-functions . The CUDA Programming Guide (like OpenCL for relaxed math) defines different error tolerance for different input ranges (see intrinsics starting from logf). Can we somehow encode this info here without resorting to worst case error tolerance?